keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,703 @@
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
+
+This creates the following complications:
+- The post-submit job must handle tags and branches for kubeadm versions
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+The cherry-pick script from k/k/hack seems to work out of the box
+for the kubeadm repository so potentially no forking will be required.
+The same cherry pick policy will apply as in k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
+
+This creates some complications:
+- Automation tools have to build kubeadm as part of the release
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Building kubeadm from tarballs
+
+For the initial proposal this will not be supported, but if there
+are plans to distribute the kubeadm source code as part of a k8s release,
+the solution would be to write a static KUBE_VERSION file in the tarball.
+
+If GitHub users download the source code directly from the kubeadm repository
+the build process must be able to detect that this is not a Git repository
+and ask the users to provide a KUBE_VERSION file.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
```

Comment:
In CI this won't work, but I don't think it will be necessary

Reply:
i can remove this sentence. not-caching the kubeadm build is not a big issue in general as it is quite fast.

Reply:
The go command now uses a build cache.

Reply:
@riking yes, problem is we are trying to share the same cache across different CI jobs.
for Bazel this works.